laurent22/joplin
# Mobile: Upgrade to React Native 0.74.1

# Summary

This pull request:
1. Applies the diff provided by the [React Native upgrade helper](https://react-native-community.github.io/upgrade-helper/?from=0.71.10&to=0.74.1&package=net.cozic.joplin&name=Joplin).
2. Migrates from `@react-native-community/clipboard` to `@react-native-clipboard/clipboard`. `@react-native-community/clipboard` doesn't build with RN 0.73 and [seems to have been renamed to `@react-native-clipboard/clipboard`](https://github.com/react-native-clipboard/clipboard/pull/87).
3. Upgrades our fork of `react-native-alarm-notification` for compatibility with newer React Native versions.

> [!NOTE]
>
> React Native 0.74 removes default support for Flipper.
>

# To-do

- [x] Upgrade our patches written for `react-native` v0.71 to work with v0.74.
    - Check whether they're still necessary?
- [x] Write a testing plan and verify that features still work on iOS and Android. (See below)

# Testing plan

> [!NOTE]
>
> Manual testing in-progress. New issues may include:
> - iOS: Pictures taken with the camera are inserted multiple times.
>    - This will need to be verified with a build of Joplin from the `dev` branch.
>

- iOS
  - [x] Sync
    1. Create a new profile
    2. Set up sync with a local copy of Joplin Server
    3. Create a notebook
    4. Enable encryption if not already enabled
    5. Add a note, attach a drawing
    6. Sync
    7. Verify that the note and drawing are visible on another client after sync.
    8. Edit the drawing on the other client and sync
    9. Sync the original mobile client and verify that the changes to the drawing are visible.
    1. Enable plugin support and install the "note tabs" and "quick links" plugins
    2. Create a new note
    3. Link to the note created while testing sync above with the "quick links" plugin
    4. Open the note viewer and click on the link
    5. Open the "note tabs" panel
    6. Close the "note tabs" panel
  - [x] Camera
    1. Open the note editor.
    2. Click on the paperclip icon in the toolbar.
    3. Click "take photo".
    4. Take a picture.
       - **Possible issue:** This attaches **two** copies of the photo to the note (with the same id). This doesn't happen if I instead click "attach photo" to attach an existing photo.
    5. Close the note editor.
    7. Verify that a picture was attached to the note.
    10. Open the note actions menu.
    11. Click "Attach"
    12. Click "Attach photo"
    13. Attach a photo.
    14. Repeat steps 7-10 for "Attach file"
  - [x] Export and import (& the share sheet)
    1. Open "Configuration" > "Import and Export"
    2. Click "Export all notes as JEX"
    3. Save the exported JEX file to the filesystem.
    4. Import the JEX file using "Import from JEX"
  - [x] Joplin Server share (tests react-native-rsa).
    1. From another Joplin Server account (with encryption enabled), share a notebook with the account the mobile device is using.
    2. Accept the share
    3. Sync
    4. Add a note to the shared notebook
    5. Sync
    6. Verify that the new note is visible on a different client.
- Android
  - [x] Sync
    1. Create a new profile
    7. Set up sync with a local copy of Joplin Server
    8. Enable encryption if not already enabled
    9. Add a note, attach a drawing
    10. Sync
    11. Verify that the note and drawing are visible on another client after sync.
    1. Enable plugin support and install the "note tabs" and "quick links" plugins
    2. Create a new note
    3. Link to the note created while testing sync above with the "quick links" plugin
    4. Open the "note tabs" panel
    5. Close the "note tabs" panel
  - [x] Camera
    1. Open the note editor.
    2. Click on the paperclip icon in the toolbar.
    3. Click "take photo".
    4. Take a picture.
    5. Close the note editor.
    6. Verify that a picture was attached to the note.
    12. Open the note actions menu.
    15. Click "Attach"
    16. Click "Attach file"
    17. Attach a file
  - [x] Export and import (& the share sheet)
    1. Open "Configuration" > "Import and Export"
    2. Click "Export all notes as JEX"
    3. Save the exported JEX file to the filesystem.
        - On Android, only sharing is implemented. To test this, I shared to Termux, which saved the exported JEX to the filesystem.
    4. Import the JEX file using "Import from JEX"
  - [x] Joplin Server share (tests react-native-rsa)
    1. From another Joplin Server account (with encryption enabled), share a notebook with the account the mobile device is using.
    2. Sync and accept the share
    3. Sync
    4. Add a note to the shared notebook
    5. Sync
  - [x] File system sync
    1. Change the sync target to "file system"
    2. Sync
    3. Verify that sync completes successfully
  - [x] Voice typing
    1. Open the note actions menu and click "Voice typing..."
    2. Wait for the language model to download.
    3. Grant permission and slowly say: "This is a test of voice typing".

Note: The encrypted share tests were re-done after a fix was merged separately.
